wagtail/documents/views/chooser.py

```diff
@@ -51,157 +67,186 @@ def get_cell_context_data(self, instance, parent_context):
-        documents = permission_policy.instances_user_has_any_permission_for(
-            request.user, ["choose"]
-        )
 
+class BaseDocumentChooseView(BaseChooseView):
+    results_template_name = "wagtaildocs/chooser/results.html"
+    filter_form_class = DocumentFilterForm
+    per_page = 10
+
+    def get_object_list(self):
+        documents = self.permission_policy.instances_user_has_any_permission_for(
+            self.request.user, ["choose"]
+        )
         # allow hooks to modify the queryset
         for hook in hooks.get_hooks("construct_document_chooser_queryset"):
-            documents = hook(documents, request)
+            documents = hook(documents, self.request)
+
+        return documents
 
-        self.q = None
-        self.is_searching = False
+    def get_filter_form(self):
+        FilterForm = self.get_filter_form_class()
+        return FilterForm(self.request.GET, collections=self.collections)
 
-        self.collection_id = request.GET.get("collection_id")
+    def filter_object_list(self, documents, form):
+        self.collection_id = form.cleaned_data.get("collection_id")
         if self.collection_id:
             documents = documents.filter(collection=self.collection_id)
-        self.documents_exist = documents.exists()
-
-        if "q" in request.GET:
-            self.searchform = SearchForm(request.GET)
-            if self.searchform.is_valid():
-                self.q = self.searchform.cleaned_data["q"]
 
-                documents = documents.search(self.q)
-                self.is_searching = True
+        self.search_query = form.cleaned_data.get("q")
+        if self.search_query:
+            documents = documents.search(self.search_query)
+            self.is_searching = True
         else:
-            self.searchform = SearchForm()
-
-        if not self.is_searching:
             documents = documents.order_by("-created_at")
 
-        paginator = Paginator(documents, per_page=10)
-        self.documents = paginator.get_page(request.GET.get("p"))
+        return documents
 
-        self.collections = permission_policy.collections_user_has_permission_for(
-            request.user, "choose"
+    @cached_property
+    def collections(self):
+        collections = self.permission_policy.collections_user_has_permission_for(
+            self.request.user, "choose"
         )
-        if len(self.collections) < 2:
-            self.collections = None
-
-        columns = [
-            TitleColumn(
-                "title",
-                label=_("Title"),
-                url_name="wagtaildocs:document_chosen",
-                link_attrs={"data-chooser-modal-choice": True},
-            ),
+        if len(collections) < 2:
+            return None
+
+        return collections
+
+    @property
+    def columns(self):
+        columns = super().columns + [
             DownloadColumn("filename", label=_("File")),
             DateColumn("created_at", label=_("Created"), width="16%"),
```

Comment:
Not related to this, so non-blocking, but looks like the changes in #8697 don't quite work in modals. Perhaps another one for #8408 ?

Reply:
Just to check - this issue has now been fixed in #8794, right?

Reply:
Yep. I tested it with this PR specifically